### ISSUE TITLE
SDK fails on Vercel serverless functions

### ISSUE THREAD

**Issue by LucilleH** [stytchauth/stytch-node#248]
Hi! I'm trying to use this library to power our auth flow in a Vercel app, and it crashes the Vercel ~~edge~~ serverless functions. I suspect it has something to do with `isomorphic-unfetch` Node polyfill.

Note that locally, or in a normal Node VM, everything works fine. Out of curiosity, since this repository is called `stytch-node`, I'm assuming that everything is geared towards Node BE. Given that, would it be possible to use `node-fetch` directly instead?

**Comment by chris-stytch** [comment]
Heya @LucilleH , thanks for reaching out! We've tested `stytch-node` with Vercel serverless functions, but I'm not sure if we've looked deeply at their Edge functions.

> ...it crashes the Vercel edge functions
Are there any error artifacts (messages, logs etc) when this happens that you could share?

**Comment by LucilleH** [comment]
Oh did I say edge functions? Sorry it was a serverless function error on fetch:
```
Unhandled Promise Rejection 
{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"TypeError: First parameter has member 'readable' that is not a ReadableStream.","reason":{"errorType":"TypeError","errorMessage":"First parameter has member 'readable' that is not a ReadableStream.","stack":["TypeError: First parameter has member 'readable' that is not a ReadableStream.","    at assertReadableStream (/var/task/node_modules/.pnpm/web-streams-polyfill@3.2.1/node_modules/web-streams-polyfill/dist/ponyfill.js:362:19)","    at convertReadableWritablePair (/var/task/node_modules/.pnpm/web-streams-polyfill@3.2.1/node_modules/web-streams-polyfill/dist/ponyfill.js:3524:9)","
at ReadableStream.pipeThrough (/var/task/node_modules/.pnpm/web-streams-polyfill@3.2.1/node_modules/web-streams-polyfill/dist/ponyfill.js:3608:29)","
at fetchFinale (node:internal/deps/undici/undici:11082:56)","
at mainFetch (node:internal/deps/undici/undici:10974:9)"...]}
Unknown application error occurred
Runtime.Unknown

```

`web-streams-polyfill` is used by `@vercel/remix`, which does the polyfill on streams. This error only occurs when `stytch-node` is used. My hypothesis is that `isomorphic-unfetch` has some side effects that changed the underlying fetch globally, which then interferes with `web-streams-polyfill`. I could be wrong.

Any help appreciated!

**Comment by chris-stytch** [comment]
Heya @LucilleH sorry about the delay here. I played around with a simple Vercel serverless function and I didn't run into the issues that you saw.

Is the repo you're working on available publicly so that I could take a look at?

**Comment by LucilleH** [comment]
@chris-stytch Hi Chris, sorry it was a private repo. But I'm pretty sure it is related to the `fetch` polyfill. I think a setup of Remix + Stytch would repro:
```
npx create-remix@latest
npm add stytch-node
// call stytch sdk and start server
npm dev
// push to vercel and run in prod
vercel deploy
```

But to save you the whole setup, I think the problem is using `isomorphic-unfetch` that switches between `unfetch` & `node-fetch` for client & server, and the dependency of [@remix-run/web-fetch](https://www.npmjs.com/package/@remix-run/web-fetch) for the remix server. 

This is not super urgent for me, as I simply re-wrote part of this library without the fetch polyfill on my end. But out of curiosity, is there a reason why `stytch-node` is not a server only package and strictly import `node-fetch`?

**Comment by chris-stytch** [comment]
@LucilleH glad you were able to get unblocked!

>  But out of curiosity, is there a reason why stytch-node is not a server only package and strictly import node-fetch?

When we last overhauled the HTTP request structure of the Node SDK, we had a few different use cases to target, e.g. Cloudflare Workers, Vercel API routes etc. Across the expected use cases, we found that our current setup worked best.

But it seems like Remix + Serverless Functions, something we didn't build explicitly for, runs into the issues that you found. I created a ticket for our team to take a look at and consider the next time we decide to crack open the request structure of the SDK.

In the meantime, feel free to join our [Slack community](https://join.slack.com/t/stytch/shared_invite/zt-nil4wo92-jApJ9Cl32cJbEd9esKkvyg) in case you need some more help, have any feedback for us, or want to show off your project!

**Comment by joshdchang** [comment]
Hi, I think I have an issue that is related to this. I'm getting the following when building my project for cloudflare workers
```
error during build:
ReferenceError: self is not defined
```
The problem seems to have to do with how Stytch is using isomorphic-unfetch. It seems like the most common polyfill used by frameworks like Nextjs, Qwik, Solid, etc. is `undici`. There is also `cross-fetch` that is popular.

This is the code from Qwik's conditional polyfill, which I'm including because it's pretty straightforward.

```
import {
  TextEncoderStream,
  TextDecoderStream,
  WritableStream,
  ReadableStream,
} from 'node:stream/web';
import { fetch, Headers, Request, Response, FormData } from 'undici';

import crypto from 'crypto';

export function patchGlobalThis() {
  if (
    typeof global !== 'undefined' &&
    typeof globalThis.fetch !== 'function' &&
    typeof process !== 'undefined' &&
    process.versions.node
  ) {
    globalThis.fetch = fetch as any;
    globalThis.Headers = Headers as any;
    globalThis.Request = Request as any;
    globalThis.Response = Response as any;
    globalThis.FormData = FormData as any;
  }
  if (typeof globalThis.TextEncoderStream === 'undefined') {
    globalThis.TextEncoderStream = TextEncoderStream;
    globalThis.TextDecoderStream = TextDecoderStream;
  }
  if (typeof globalThis.WritableStream === 'undefined') {
    globalThis.WritableStream = WritableStream as any;
    globalThis.ReadableStream = ReadableStream as any;
  }
  if (typeof globalThis.crypto === 'undefined') {
    globalThis.crypto = crypto.webcrypto as any;
  }
}
```

**Comment by joshdchang** [comment]
Since `fetch` has been built into Node for a while now, commenting out these two lines in [stytch-node/lib/shared
/index.ts](https://github.com/stytchauth/stytch-node/blob/87b495ede1d606d02319482d957dd816de987f1f/lib/shared/index.ts) actually fixed my issue, so I'm confident the error I'm getting is an issue with `isomorphic-unfetch`.

```ts
// import * as fetchImport from "isomorphic-unfetch";

// const fetch = (fetchImport.default || fetchImport) as typeof fetchImport.default;
```

This should be pretty straightforward to fix. Just install `undici` and replace those two lines with the following:
```ts
import { fetch } from 'undici';

if (typeof global !== 'undefined' && typeof globalThis.fetch !== 'function' && typeof process !== 'undefined' && process.versions.node) {
  globalThis.fetch = fetch;
}
```

**Comment by max-stytch** [comment]
Hey @joshdchang thanks for the pointer to `undici`! Here are my thoughts so far:
- We still support Node v16 LTS, which doesn't have native fetch unless ran with the [--experimental-fetch](https://nodejs.org/en/blog/release/v16.15.0) flag, so we still need a fallback
- We prefer ponyfills over polyfills, because ideally our library should not be mucking around with any other libraries. If installing `stytch-node` breaks something else, that's a bad time
- I was unable to reproduce the `ReferenceError: self is not defined` using the latest version of wrangler. I've opened up https://github.com/stytchauth/stytch-node/pull/252 to show how I am testing with Cloudflare Workers.
- I was able to replicate the `"TypeError: First parameter has member 'readable' that is not a ReadableStream." so we should be able to take a look at that soon

**Comment by max-stytch** [comment]
!! Looks like Node 16 got [EOL'd early!](https://nodejs.org/en/blog/announcements/nodejs16-eol) Looks like we can do something sooner.

**Comment by max-stytch** [comment]
https://github.com/stytchauth/stytch-node/pull/255 should address this by dropping support for Node 16, dropping `isomorphic-unfetch` completely, and using the builtin `fetch` API directly. We'll likely take a few days to let this bake, but I've tested it out in Cloudflare Workers, Vercel Remix Serverless, Vercel NextJS Serverless, & Vercel NextJS Edge Runtimes so far.

If you'd like to try it out yourself, a package is available at [`stytch@9.0.0-rc.0`](https://www.npmjs.com/package/stytch/v/9.0.0-rc.0) or `stytch@next`.
